Merging "development" branch into "master" branch

Elliott-Marples opened the pull request.

Elliott-Marples committed:
Initialised game

Initialised game with a room and player objects.

Elliott-Marples committed:
Expanded player & room classes

Ensures correct player input & added directions for rooms.

Elliott-Marples committed:
Added room navigation

Rooms can now be navigated between easily with cardinal directions using a 2d array

Elliott-Marples committed:
Added descriptions to rooms

Added in depth to each room that the player can enter.

Elliott-Marples committed:
Added appropriate comments to code

Elliott-Marples committed:
Added new lines to separate console lines

Elliott-Marples committed:
Added items

Added items, the ability to pick items up and use them. Also added room inaccessibility, attack and happiness properties for the player and fixed some variable names.

Elliott-Marples committed:
Removed console line from program

RavenNight5 commented on Player.cs:
```diff
@@ -1,25 +1,209 @@
+            set
+            {
+                // If the value is null or empty an appropriate message is printed
+                if (string.IsNullOrEmpty(value))
+                {
+                    Console.WriteLine("\nName cannot be empty.");
+                }
```
Here you could include another form of validation to check the input name does not contain any whitespaces or numbers. An example of the if statement could be:

```
if (string.IsNullOrEmpty(value))
{
    Console.WriteLine("\nName cannot be empty.");
} 
else if (name.Any(char.IsWhiteSpace) || !name.All(char.IsLetter))
{
    Console.WriteLine("\nName cannot contain whitespace or numbers.");
}
```

RavenNight5 commented on Player.cs:
```diff
@@ -1,25 +1,209 @@
+                // If the value is less than 1 an appropriate message is printed
+                if (value < 1)
+                {
+                    Console.WriteLine("\nHappiness cannot be zero or negative.");
```
Considering this would be an error that shouldn't occur, having an if statement to double check is good, but this checking can easily be done in one line with the ```Debug.Assert()``` method.

E.g. ```Debug.Assert(!(value < 1), "Error: Happiness cannot be zero or negative.")  // Message is brought up in a separate debug window that can be ignored by the user, allowing the program to continue running.```

When ```value``` is greater than 0 no error will be thrown and the program can continue. Therefore this one line can replace the whole if statement.

This holds true for any other similar statements (such as the integer Attack). However, I would not use this for health as it would disrupt the gameplay when the player dies since the health may naturally be updated to less than 1.

RavenNight5 reviewed (commented):
Finished Review! Nice use of the matrix

LilahRet commented on Player.cs:
```diff
@@ -1,25 +1,209 @@
+            if (_equippedItem != null)
+            {
+                // The equipped item is added back to the inventory
+                _inventory.Add(_equippedItem);
```
You should add a message saying you have unequipped an item after doing so

LilahRet commented on Game.cs:
```diff
@@ -1,25 +1,368 @@
+            while (player == null || player.Name == null)
+            {
+                // Asks user for their name and sets player's health
+                Console.Write("Enter your name:\n> ");
```
Make it so you can't enter numbers or spaces for a name

LilahRet commented on Game.cs:
```diff
@@ -1,25 +1,368 @@
+            Room room5 = new Room("a chamber with doors to the east and west.\nThere's a closed metal gate to the south.\nIn front of the gate is a stone on a pedestal.", item: "Stone");
+            Room finalRoom = new Room("a vast underground arena with a massive slime creature in the centre.", isAccessible: false);
+            Room exit = new Room("Exit.");
+
```
Make it so that when you have already searched a room before, it will display a message like "You have already searched this room" to help with navigation

LilahRet reviewed (commented):
Nice code, well made and implemented. Can be a little hard to navigate through all the rooms.

Elliott-Marples committed:
Added changes specified in code reviews

Added more name checks, used `debug.assert` and a visited room property that checks whether a room has been previously visited by the player.
